tcp2: Implement a blocking connect

ozhuraki opened the pull request:
In order to implement a blocking connect, add a semaphore
and block on it in net_tcp_connect().

The semaphore is released when ESTABLISHED state is reached.

In case tcp_conn_unref() is called while waiting on the semaphore,
defer the unreference, tcp_conn_unref() will be called from
net_tcp_connect().

Fixes #26695

ozhuraki committed:
net: tcp2: Refactor net_tcp_connect()

In order to improve readability, refactor and simplify
the control flow in net_tcp_connect().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Send FIN only in ESTABLISHED state

net_tcp_put() can be called before ESTABLISHED state
is reached, send FIN only in ESTABLISHED state.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Check for SYN and ACK in SYN_SENT

In order for events to be correctly interpreted in SYN_SENT,
check for SYN and ACK simultaneosly.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Implement a blocking connect

In order to implement a blocking connect, add a semaphore
and block on it in net_tcp_connect().

The semaphore is released when ESTABLISHED state is reached.

In case tcp_conn_unref() is called while waiting on the semaphore,
defer the unreference, tcp_conn_unref() will be called from
net_tcp_connect().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

jukkar reviewed (commented):
This certainly works for current stack. But it won't work if we get the user mode networking done because there we would need to separate the userspace application by IPC from core stack. Of course we are not yet there, just pondering should make the code so that it would be easier to split the stuff.

Edit: lets go with this solution and figure out what needs to be done with usermode networking later. The plans are still so vague that it is difficult to know what direction to take. Can you fix the compile warning so that sanity is happy.

ozhuraki committed:
tests: net: tcp2: Update timeout values

Since the connect is now blocking, update the timeout values
for the tests to pass.

In case of resend test, set the timeout to 1000ms to account
for the retransmission.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
tests: net: tcp2: Update test_sem_take()

Fail solely on semaphore timeout.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
tests: net: tcp2: Check for failed connect on SYN resend

Since the peer isn't replying on SYN resend, update the
test to check for failed connect.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -1129,20 +1129,19 @@ static void tcp_in(struct tcp *conn, struct net_pkt *pkt)
-					      NET_CONTEXT_CONNECTED);
-			if (FL(&fl, &, PSH)) {
+			conn_ack(conn, th_seq(th) + 1);
+			if (len) {
```
have an empty line before the if (since previous line it not doing anything with len)

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -1458,6 +1469,17 @@ int net_tcp_connect(struct net_context *context,
 
+#if !defined(CONFIG_NET_TEST_PROTOCOL)
+	conn->in_connect = true;
+	if (k_sem_take(&conn->connect_sem, timeout) != 0 &&
```
empty line before if

sjg20 commented on subsys/net/ip/tcp2.c:
```diff
@@ -280,11 +280,20 @@ static void tcp_send_queue_flush(struct tcp *conn)
 
 	NET_DBG("conn: %p, ref_count=%d", conn, ref_count);
 
+#if !defined(CONFIG_NET_TEST_PROTOCOL)
```
if() instead of #if ?

sjg20 commented on subsys/net/ip/tcp2_priv.h:
```diff
@@ -201,6 +201,8 @@ struct tcp { /* TCP connection */
 	size_t send_retries;
 	struct k_delayed_work timewait_timer;
 	struct net_if *iface;
+	struct k_sem connect_sem;
```
comments?

mniestroj commented:
Why don't we use `cb` parameter of `net_tcp_connect()` (both in TCP and TCP2) to notify about established connection? What we are really missing in #26695 is that posix `connect()` should block by default, not necessarily `net_tcp_connect()`. Such behavior could be implemented by waiting on semaphore in `zsock_connect_ctx()`.

In Linux userspace there is also a possibility to change socket file descriptor to non-blocking mode. Then you can call `connect()`, which won't block. `poll()` can be used on such file descriptor to get notification about connection being established. This gives userspace a mechanism for timeout, as file descriptor can be closed when userspace decides it takes too long to connect. If at some point we would like such mechanism in Zephyr, then it would be much easier to put all the logic in the `zsock_connect_ctx()` and decide whether to use semaphore (blocking mode) or signal `poll()` in `cb` implementation passed to `net_context_connect()`.

johngrey-dev commented:
This will be the better solution, but for now the module is unusable. 
Temporarily solving it as currently suggested will be extremely helpful.

Thank you!

> Why don't we use `cb` parameter of `net_tcp_connect()` (both in TCP and TCP2) to notify about established connection? What we are really missing in #26695 is that posix `connect()` should block by default, not necessarily `net_tcp_connect()`. Such behavior could be implemented by waiting on semaphore in `zsock_connect_ctx()`.
> 
> In Linux userspace there is also a possibility to change socket file descriptor to non-blocking mode. Then you can call `connect()`, which won't block. `poll()` can be used on such file descriptor to get notification about connection being established. This gives userspace a mechanism for timeout, as file descriptor can be closed when userspace decides it takes too long to connect. If at some point we would like such mechanism in Zephyr, then it would be much easier to put all the logic in the `zsock_connect_ctx()` and decide whether to use semaphore (blocking mode) or signal `poll()` in `cb` implementation passed to `net_context_connect()`.

jukkar commented:
> Why don't we use `cb` parameter of `net_tcp_connect()` (both in TCP and TCP2) to notify about established connection? What we are really missing in #26695 is that posix `connect()` should block by default, not necessarily `net_tcp_connect()`. Such behavior could be implemented by waiting on semaphore in `zsock_connect_ctx()`.

Indeed, that sounds actually a better solution. I think we need to revisit this fix anyway if I manage to figure out how to implement the user mode networking support. As we are understaffed atm and this fix works, I am merging this as is. Patches are welcome to implement this differently.
